File: String/Palindrome/Manacher.hs
import Data.Array (assocs, listArray, (!))
import Test.QuickCheck

palindrome :: Eq a => [a] -> [(Int, Int, Int)]
palindrome str = [(l, r, go l r) | l <- [1 .. n], r <- [l, l + 1], r <= n]
  where
    n = length str
    a = listArray (1, n) str
    go l r
      | l < 1 || r > n  = 0
      | a!l /= a!r      = 0
      | otherwise       = succ $ go (pred l) (succ r)

manacher :: Eq a => [a] -> [(Int, Int, Int)]
manacher str = [(div i 2, div (i + 1) 2, e) | (i, e) <- assocs b]
  where
    n = length str
    a = listArray (1, n) str
    pal l r
      | l < 1 || r > n = 0
      | a!l /= a!r     = 0
      | otherwise      = succ $ pal (pred l) (succ r)
    b = listArray (2, 2*n) $ go 2 (0, 0)
    go k (e, c) = if l > n then [] else t: go (k + 1) (e', c')
      where
        l = div k 2
        r = k - l
        s = if k <= e then min (b!(2*c-k)) $ 1 + div (e-k) 2 else 0
        t = s + pal (l - s) (r + s)
        (e', c') = max (e, c) (2 * (r + t - 1), k)

prob_palindrome_model :: Eq a => [a] -> Bool
prob_palindrome_model s = manacher s == palindrome s

data Color = Red | Green | Blue deriving (Eq, Show)
instance Arbitrary Color where
  arbitrary = elements $ concatMap (uncurry replicate) $
    [(100, Red), (10, Green), (1, Blue)]

main :: IO ()
main = do
  check (prob_palindrome_model :: [Bool] -> Bool)
  check (prob_palindrome_model :: [Color] -> Bool)
  check (prob_palindrome_model :: [Char] -> Bool)
  where
    check :: Testable prop => prop -> IO ()
    check = quickCheckWith stdArgs{maxSuccess=1000}

File: String/Palindrome/Manacher.cpp
#include <algorithm>

using namespace std;

struct Manacher {
  vector<int> p;

  template<typename T>
  void init(int n, const T s[]) {
    p.resize(max(0, 2 * n - 1));
    for (int i = 0, j = 0, k = 0; i <= 2 * (n - 1); ++i) {
      int d = i < k ? min(p[j + j - i], (k - i) / 2) : 0;
      int a = i / 2 - d, b = (i + 1) / 2 + d;
      while (0 <= a && b < n && s[a] == s[b]) {
        --a;
        ++b;
        ++d;
      }
      p[i] = d;
      if (k < 2 * b - 1) {
        j = i;
        k = 2 * b - 1;
      }
    }
  }
};

// brute-force
struct Palindrome {
  vector<int> p;

  template<typename T>
  void init(int n, const T s[]) {
    p.resize(max(0, 2 * n - 1));
    for (int i = 0; i < (int)p.size(); ++i) {
      int a = i / 2, b = (i + 1) / 2, d = 0;
      while (0 <= a && b < n && s[a] == s[b]) {
        --a;
        ++b;
        ++d;
      }
      p[i] = d;
    }
  }
};
